program/src/processor.rs

```diff
@@ -351,6 +355,82 @@ pub fn process_stake_deposit(
     lido.save(accounts.lido)
 }
 
+/// Unstakes from a validator, the funds are moved to the stake defined by the
+/// validator's unstake seed. Caller must be a maintainer.
+pub fn process_unstake(
+    program_id: &Pubkey,
+    amount: Lamports,
+    raw_accounts: &[AccountInfo],
+) -> ProgramResult {
+    let accounts = UnstakeAccountsInfo::try_from_slice(raw_accounts)?;
+    let mut lido = deserialize_lido(program_id, accounts.lido)?;
+    lido.check_maintainer(accounts.maintainer)?;
+    lido.check_stake_authority(program_id, accounts.lido.key, accounts.stake_authority)?;
+    let destination_bump_seed = check_unstake_accounts(program_id, &lido, &accounts)?;
+
+    let validator = lido.validators.get(accounts.validator_vote_account.key)?;
+    let seeds = [
+        &accounts.lido.key.to_bytes(),
+        &accounts.validator_vote_account.key.to_bytes(),
+        VALIDATOR_UNSTAKE_ACCOUNT,
+        &validator.entry.unstake_seeds.end.to_le_bytes()[..],
+        &[destination_bump_seed],
+    ];
+
+    split_stake_account(
+        accounts.lido.key,
+        &lido,
+        &SplitStakeAccounts {
+            source_stake_account: accounts.source_stake_account,
+            destination_stake_account: accounts.destination_stake_account,
+            authority: accounts.stake_authority,
+            system_program: accounts.system_program,
+            stake_program: accounts.stake_program,
+        },
+        amount,
+        &[&seeds],
+    )?;
```

Comment:
Ok so now we test that you need to leave at last `MINIMUM_STAKE_ACCOUNT_BALANCE`, but with that we can’t do validator removal.

We can add a special case for unstaking exactly the current balance. That should be allowed, and also bump the `begin` seed, as it closes the source stake account (we should confirm that in a test too).

Alternatively we can remove the check that we leave the `MINIMUM_STAKE_ACCOUNT_BALANCE` in the account, but then we can end up with stake accounts that have less than the minimum, which would be very confusing, so I think special-casing unstaking 100% is fine. Also because in that case we need to bump the `begin` seed and in other cases we don’t.

Reply:
I think we should allow unstaking > `MINIMUM_STAKE_ACCOUNT_BALANCE` if the validator is being deactivated. That should be fine

Reply:
> I think we should allow unstaking > MINIMUM_STAKE_ACCOUNT_BALANCE if the validator is being deactivated. That should be fine

Ah, we only allow it for inactive validators? That makes sense. We could go a step further and require that if you unstake from an inactive validator, you unstake the full account.

Reply:
Implemented it like that.